From 9aae5cf478fe3c04cb956b23b166fd49261b8d61 Mon Sep 17 00:00:00 2001
From: Giovanni Campagna <gcampagna@src.gnome.org>
Date: Sat, 29 Oct 2011 14:35:09 +0200
Subject: [PATCH 3/3] NetworkMenu: fix logic for updating wifi icon

Previously, we connected to notify::strength only if there was
already a signal connected, and the AP changed (thus, by induction,
we never connected). As a result, the icon became stale and different
from that shown inside the menu (which is correctly updated).

https://bugzilla.gnome.org/show_bug.cgi?id=650007
---
 js/ui/status/network.js |   12 +++++++-----
 1 files changed, 7 insertions(+), 5 deletions(-)

diff --git a/js/ui/status/network.js b/js/ui/status/network.js
index aa9b277..d40f7a6 100644
--- a/js/ui/status/network.js
+++ b/js/ui/status/network.js
@@ -2061,10 +2061,11 @@ NMApplet.prototype = {
                         }
                         this.setIcon('network-wireless-connected');
                     } else {
-                        if (this._accessPointUpdateId && this._activeAccessPoint != ap) {
-                            this._activeAccessPoint.disconnect(this._accessPointUpdateId);
+                        if (this._activeAccessPoint != ap) {
+                            if (this._accessPointUpdateId)
+                                this._activeAccessPoint.disconnect(this._accessPointUpdateId);
                             this._activeAccessPoint = ap;
-                            this._activeAccessPointUpdateId = ap.connect('notify::strength', Lang.bind(function() {
+                            this._activeAccessPointUpdateId = ap.connect('notify::strength', Lang.bind(this, function() {
                                 this.setIcon('network-wireless-signal-' + signalToIcon(ap.strength));
                             }));
                         }
@@ -2091,8 +2092,9 @@ NMApplet.prototype = {
                     break;
                 }
 
-                if (this._mobileUpdateId && this._mobileUpdateDevice != dev) {
-                    this._mobileUpdateDevice.disconnect(this._mobileUpdateId);
+                if (dev.mobileDevice != this._mobileUpdateDevice) {
+                    if (this._mobileUpdateId)
+                        this._mobileUpdateDevice.disconnect(this._mobileUpdateId);
                     this._mobileUpdateDevice = dev.mobileDevice;
                     this._mobileUpdateId = dev.mobileDevice.connect('notify::signal-quality', Lang.bind(this, function() {
                         this.setIcon('network-cellular-signal-' + signalToIcon(dev.mobileDevice.signal_quality));
-- 
1.7.7.3

